HAWKULAR-844

jsanda opened the pull request.

ppalaga committed:
Upgrade to hawkular-parent 26

ppalaga committed:
Replace org.jboss.security.util.StringPropertyReplacer with
org.jboss.common.beans.property.PropertiesValueResolver

ppalaga committed:
Migrate to ActiveMQ Artemis

Commit:
[HAWKULAR-844] initial changes to remove ActiveMQ dependencies

Commit:
[HAKWULAR-844] some broken itests

Initially I was putting tests in a separate module so that I could test the bus i
as a JBoss module. That is when I discovered CDI injection does not work with
JBoss modules, at least with what I have done. At least for now the tests can
live in the hawkular-bus-common module and be executed in the test phase which
makes for easier and faster test execution. The Bus class is a just a prototype.

ppalaga committed:
Feature pack

ppalaga committed:
Integration tests - some polishing stil needed

Commit:
[HAWKULAR-844] work around for class name header

We set a header on each message for the class name which is resulting in the
following exception with Artemis,

javax.ejb.EJBException: javax.jms.JMSRuntimeException: AMQ129012: The property name 'org.hawkular.bus.common.BasicMessage.className' is not a valid java identifier.
    Caused by: javax.jms.JMSRuntimeException: AMQ129012: The property name 'org.hawkular.bus.common.BasicMessage.className' is not a valid java identifier."}}

At least as a temporary work around I have changed the header name to a
recognized, supported name.

Commit:
[HAWKULAR-844] fix Session.createQueue, Session.createTopic calls

Using the JNDI path does not work with Artemis as it did with ActiveMQ.

Commit:
[HAWKULAR-844] a couple minor fixes for module deployment

We need a dependency on the jackson annotation module. I discovered that during
manual testing. I also modified the build slightly to include the assembly id
in the zip artifact name because the main hawkular build expects this and will
not lay down the module properly if the assembly id is not included. Lastly, I
have removed constructors that were used specifically with ActiveMQ.

Commit:
[HAWKULAR-844] undo changes made to restore behavior

Commit:
[HAWKULAR-844] clean up after rebase

Commit:
[HAWKULAR-844] fix test failures

Commit:
[HAWKULAR-844] update travis config and delete scratch test code

Now that we execute integration tests in a separate module, we need to run the
install phase of maven during travis builds. The BusTest class that is deleted
was previously deleted but accidentially brought back after merge/rebase.

ppalaga commented:
@jsanda I vote for removing this commented-out dependency unless there some reason to keep it: https://github.com/jsanda/hawkular-bus/blob/31aa5638a1f64cd1fcc8cd454567805cf0f6100e/hawkular-bus-common/pom.xml#L102-L106

Commit:
[HAWKULAR-844] remove/clean up commented out code

ppalaga commented:
@jsanda given that you adopted my feature pack work, we can safely remove this old module packaging, can't we? https://github.com/jsanda/hawkular-bus/blob/31aa5638a1f64cd1fcc8cd454567805cf0f6100e/hawkular-bus-common/pom.xml#L115-L148
https://github.com/jsanda/hawkular-bus/tree/31aa5638a1f64cd1fcc8cd454567805cf0f6100e/hawkular-bus-common/src/main/resources
and https://github.com/jsanda/hawkular-bus/tree/31aa5638a1f64cd1fcc8cd454567805cf0f6100e/hawkular-bus-common/src/main/scripts

jsanda commented:
> I vote for removing this commented-out dependency unless there some reason to keep it: https://github.com/jsanda/hawkular-bus/blob/31aa5638a1f64cd1fcc8cd454567805cf0f6100e/hawkular-bus-common/pom.xml#L102-L106

Done

ppalaga commented:
I see `hawkular-bus-common/src/main/resources/.DS_Store` added. Should it rather not be ignored?

jsanda commented:
Yes, I guess we should remove the older module assembly code; however, I would like to do the same manual smoke testing with the feature pack that I have been doing with the module. I will do that and provided there are non regressions I will remove the old module packaging.

jsanda commented:
> I see hawkular-bus-common/src/main/resources/.DS_Store added. Should it rather not be ignored?

Yeah, I will remove and update my git config.

Commit:
remove .DS_Store file from source control

ppalaga commented:
> Yes, I guess we should remove the older module assembly code; however, I would like to do the same manual smoke testing with the feature pack that I have been doing with the module. I will do that and provided there are non regressions I will remove the old module packaging.

I am OK with not removing the old packaing stuff in this PR, but we should not forget. It would perhaps be handy to mark/comment all the named resouces that they can/will be removed once the condition X is fulfilled.

ppalaga commented:
@jsanda If you'd mark the old packaging recources somehow, I'd be ready to merge the PR.

Commit:
[HAWKULAR-844] add TODO about feature packs

ppalaga commented:
@jsanda perfect! Good to merge from my side.

jmazzitelli commented on hawkular-bus-common/src/main/java/org/hawkular/bus/common/Bus.java:
```diff
@@ -0,0 +1,58 @@
+ * @author jsanda
+ */
+@ApplicationScoped
+public class Bus {
```
what is this class all about? send() is empty

jsanda replied:
I initially created it as part of some refactoring yet to be proposed/discussed. That refactoring effort though is largely centered around adopting JMS 2.0 APIs. The purpose of this PR though is only to move off of ActiveMQ and onto the default messaging subsystem in WF 10 (i.e, Artemis). On that basis, I am perfectly fine with removing the class.

jmazzitelli commented on hawkular-bus-common/src/main/java/org/hawkular/bus/common/MessageProcessor.java:
```diff
@@ -338,42 +339,6 @@ public MessageId sendWithBinaryData(ProducerConnectionContext context, BasicMess
     /**
-     * Same as {@link #sendRPC(ProducerConnectionContext, BasicMessage, Class, Map)} with <code>null</code> headers.
-     */
-    public <R extends BasicMessage> ListenableFuture<BasicMessageWithExtraData<R>> sendRPC(
```
so are we getting rid of the RPC-like (synchronous) messaging entirely from the hawkular bus API?

jsanda replied:
I removed that method because I did not see it used anywhere in any of the of the component repos. I only found it used in a one of the samples and in a test method.

jmazzitelli replied:
I think no one is going to use that. rather than just have things uncommented out everywhere, just remove this stuff entirely - the tests, samples, API. I thought some folks might want the ability to wait for a message to be processed and results returned (a-la RPC) but no one seems to want that.

jmazzitelli commented on hawkular-bus-common/src/main/java/org/hawkular/bus/common/MessageProcessor.java:
```diff
@@ -487,26 +456,14 @@ protected Message createMessageWithBinaryData(ConnectionContext context, BasicMe
         // we are using a ActiveMQ-specific feature that allows us to stream blobs
```
this three-line comment can go away - looks like the new session API provides what ActiveMQ did not.

jmazzitelli commented on hawkular-bus-common/src/main/java/org/hawkular/bus/common/SimpleMessageProcessor.java:
```diff
@@ -67,16 +65,16 @@ public MessageId send(BasicMessage basicMessage, Map<String, String> headers) th
-            throws JMSException {
-        return super.sendRPC(producerCtx, basicMessage, expectedResponseMessageClass, headers);
-    }
+//    public <R extends BasicMessage> ListenableFuture<BasicMessageWithExtraData<R>> sendRPC(BasicMessage basicMessage,
```
if we no longer want to support synchronous messaging (I actually thought it was a neat feature of the hawkular bus API, but I don't think anyone is using it today) then delete this entirely.

jsanda replied:
The motivation for using ListenableFuture was to support async messaging. I removed it for two reasons. It was not used anywhere. And in a separate PR/ticket I am going to propose a different API for async that utilizes RxJava.

jmazzitelli commented on hawkular-bus-common/src/main/java/org/hawkular/bus/common/consumer/AbstractBasicMessageListener.java:
```diff
@@ -131,49 +130,31 @@ public void setConsumerConnectionContext(ConsumerConnectionContext consumerConne
-                basicMessageClass = (Class<T>) Class.forName(basicMessageClassName, true, cl);
+            String basicMessageClassName =
+                    (String) message.getObjectProperty(MessageProcessor.HEADER_BASIC_MESSAGE_CLASS);
+            log.infof("About to parse message of type [%s]", basicMessageClassName);
```
make these tracef or debugf - this is going to be noisy

jmazzitelli commented on hawkular-bus-common/src/main/java/org/hawkular/bus/common/consumer/AbstractBasicMessageListener.java:
```diff
@@ -131,49 +130,31 @@ public void setConsumerConnectionContext(ConsumerConnectionContext consumerConne
-            // unless a subclass wants to substitute another class for it.
-            String basicMessageClassName = message.getStringProperty(MessageProcessor.HEADER_BASIC_MESSAGE_CLASS);
-            if (basicMessageClassName != null) {
-                String desired = convertReceivedMessageClassNameToDesiredMessageClassName(basicMessageClassName);
```
you gotta put this back - me and jay purposefully made these changes to support things alerts/events needs. This change reverts back to old code that will now break alerts/events.

jmazzitelli commented on hawkular-bus-common/src/main/java/org/hawkular/bus/common/consumer/BasicMessageListener.java:
```diff
@@ -50,6 +52,9 @@ protected BasicMessageListener(ClassLoader basicMessageClassLoader) {
     @Override
     public void onMessage(Message message) {
+
+        log.infof("Received raw message [%s]", message);
```
change to debugf - this is going to be noisy

jmazzitelli commented on hawkular-bus-common/src/main/java/org/hawkular/bus/common/consumer/BytesMessageInputStream.java:
```diff
@@ -0,0 +1,70 @@
+ * @author <a href="https://github.com/ppalaga">Peter Palaga</a>
+ *
+ */
+public class BytesMessageInputStream extends InputStream {
```
I think you need to override more methods in there. close() for example - when this stream is closed, don't you want to close the delegate?

ppalaga replied:
The delegate is a `BytesMessage` - I do not see that it can be closed in any way. Can you?

jsanda replied:
> when this stream is closed, don't you want to close the delegate?

`delegate` is not a stream. It is a BytesMessage.

jmazzitelli replied:
Ah, I didn't pay attention. OK, BytesMessage isn't really a stream then never mind.

Commit:
[HAWKULAR-844] remove dead code and obsolete comments

jmazzitelli commented on hawkular-bus-test-common/src/test/java/org/hawkular/bus/common/test/RPCTest.java:
```diff
@@ -16,383 +16,359 @@
 /**
  * Tests request-response messaging.
  */
 public class RPCTest {
```
if we are getting rid of the sendRPC APIs, then just get rid of this entirely.

Commit:
[HAWKULAR-844] lower logging to debug

Commit:
[HAWKULAR-844] revert behavior for determining message body class name

Removing unused test class. I think there are more classes in
hawkular-bus-test-common that can be removed. I can address that in some follow
up commits after I have had time to further review component integration tests.

jsanda commented:
Are there any other changes that should be made? If not, @ppalaga would you mind merging?

jmazzitelli commented:
correct me if I'm wrong - but we don't want to merge this into master until we get on WF10. otherwise, we can't release bus anymore until we move everything over to WF10

jsanda commented:
We are doing this as part of the WF 10 upgrade. I have another PR in the hawkular repo (https://github.com/hawkular/hawkular/pull/690) that updates the hawkular server build. It cannot be merged until this PR is merged.
